source/common/protobuf/utility.cc

```diff
@@ -505,6 +506,79 @@ std::string MessageUtil::CodeEnumToString(ProtobufUtil::error::Code code) {
   }
 }
 
+namespace {
+
+// Recursive helper method for MessageUtil::redact() below.
+void redactInPlace(Protobuf::Message* message, bool ancestor_is_redacted) {
+  // If the message is an `Any`, we can't use any type-based introspection. Instead, we have to
+  // unpack it to its original type to redact it...
+  auto* any = dynamic_cast<ProtobufWkt::Any*>(message);
+  if (any != nullptr) {
+    // 1. Extract the type name from the URL. This is guaranteed to be the portion of the URL
+    // following the last '/' character.
+    const std::string& type_url = any->type_url();
+    const std::size_t last_slash = type_url.find_last_of('/');
+    const std::string type_name = type_url.substr(last_slash + 1);
+    const auto* descriptor =
+        Protobuf::DescriptorPool::generated_pool()->FindMessageTypeByName(type_name);
+
+    // 2. Instantiate a mutable message of the correct dynamic type, and unpack the `Any` into it.
+    Protobuf::DynamicMessageFactory dmf;
+    std::unique_ptr<Protobuf::Message> typed_message(dmf.GetPrototype(descriptor)->New());
+    any->UnpackTo(typed_message.get());
+
+    // 3. Redact the typed message and repack it into the original `Any`.
+    redactInPlace(typed_message.get(), ancestor_is_redacted);
+    any->PackFrom(*typed_message);
+
+    return;
+  }
+
+  // Otherwise, use reflection to traverse all populated fields of this message...
```

Comment:
The other concern is that this might be applied to a `Struct` encoded proto that is later run through `jsonConvert` and turned into a real proto. You don't have sufficient type information when it's a `Struct`.

This largely depends on where we call this. It wouldn't be safe to call it on the bootstrap for example.

We probably don't need to boil the ocean in this PR, but we should carefully document the limits on where this utility can be safely used.

Reply:
Yeah, with `Any`, at least we have `type_url` to reconstruct type info from. We're out of luck with `Struct`. So yeah, I think you're right, I should document that limitation.

Reply:
Let's take bootstrap as an example; are we confident that early logging of bootstrap is not going to leave unredacted Structs in debug logs?